src/modules/python/pbs_hooks/PBS_alps_inventory_check.PY

```diff
@@ -0,0 +1,351 @@
+    6) A compute node that is offlined in PBS, but still preset as a
+       batch node in ALPS, would cause a false inventory mismatch
+       since only "up" nodes were included in the pbs_nids_set. This
+       would in turn cause the hook to HUP inventory MoM every cycle.
+       It's perfectly reasonable for an Admin to offline a node
+       without necessarily downing it in ALPS. We deal with this by:
+       a) treating node states as a bitmask rather than as a set of
+          integer values
+       b) distinguishing between "down" states ("down", "unknown",
+          "stale") and offline
+       c) excluding down or offline login nodes from consideration as
+          the inventory MoM
+       d) special-casing offline compute nodes by removing their nids
+          from the apstat_nids_set if present and excluding them from
+          the pbs_nids_set. This essentially means that we don't care
+          whether an offline node is present in ALPS or not.
+    7) When we check to see whether the current node matches the
+       inventory MoM, we actually resolve the inventory MoM and see
+       if its IP matches a valid IP from the current node . This allows
+       the hook to continue to work even if a site changes the default
+       hostname of a MoM node after installation.
+"""
+
+import pbs
+import time
+import sys
+import os
+import re
+import socket
+from signal import SIGHUP
+
+XTHOSTNAME = "/etc/xthostname"
+APSTAT_CMD = "/opt/cray/alps/default/bin/apstat"
+SIOCGIFADDR = 0x8915
+
+
+def get_mom_home():
+    "Return the path to the PBS home directory"
+
+    for v in ("PBS_MOM_HOME", "PBS_HOME"):
+        if v in os.environ:
+            return os.environ[v]
+
+    home = None
+    conf_file = os.environ.get("PBS_CONF_FILE", "/etc/pbs.conf")
+    conf = open(conf_file, "r")
+    for line in conf:
+        (key, val) = line.strip().split("=")
+        if key == "PBS_MOM_HOME":
+            conf.close()
+            home = val
+            # PBS_MOM_HOME takes priority over PBS_HOME, so we are done
+            # searching
+            break
+        elif key == "PBS_HOME":
+            home = val
+    conf.close()
+
+    return home
+
+
+def get_apstat_nids(msg):
+    """
+    Returns the number of nodes reported by ALPS as marked "up" and of type
+    "batch".
+
+    Sample output of the command 'apstat -rn':
+
+    NID Arch State HW Rv Pl  PgSz     Avl    Conf Placed PEs Apids
+    2   XT UP  B 16  -  -    4K 4194304       0      0   0
+    3   XT UP  B 16  -  -    4K 4194304       0      0   0
+    """
+
+    if not os.path.isfile(APSTAT_CMD):
+        msg += ["ALPS Inventory Check: apstat command can not be found at %s" %
+                (APSTAT_CMD)]
+        return None
+
+    apstat_nids = set()
+    cmd_apstat = APSTAT_CMD + " -nv"
+
+    # Using os.popen instead of subprocess.Popen because pbs_python 2.5.1 does
+    # not ship the subprocess module
+    apstat_out = os.popen(cmd_apstat)

```

Comment:
You shouldn't need to worry about that.  You're merging into master which uses python 2.7.